torch/distributed/fsdp/_optim_utils.py

```diff
@@ -181,25 +186,24 @@ def _communicate_optim_state(
                 fsdp_module.world_size == 1
                 or fsdp_module.sharding_strategy == FSDP.ShardingStrategy.NO_SHARD
             ):
-                tensor_state[state_name] = value.cpu()
+                tensor_state[state_name] = value
                 continue
             if not value.is_cuda:
                 value = value.to(fsdp_module.compute_device)
-            if tensor_buffer is None:
-                # Assume that positive-dimension tensor optimizer state
-                # has the same shape as the sharded flattened parameter
-                buffer_size = flat_param._full_param_padded.size()  # type: ignore[attr-defined]
-                tensor_buffer = value.new_zeros(*buffer_size)
+            # Assume that positive-dimension tensor optimizer state
+            # has the same shape as the sharded flattened parameter
+            buffer_size = flat_param._full_param_padded.size()  # type: ignore[attr-defined]
+            tensor_buffer = value.new_zeros(*buffer_size)
```

Comment:
Curious why we remove the reuse of `tensor_buffer`? For Adam, this allocates `tensor_buffer` twice per parameter instead of once?

Reply:
We can reuse the tensor_buffer because the original code move it to CPU immediately. This PR delay cpu() call, so we have to allocate the buffer twice. It does not add too much cost for the model I tested.